modules/database/database_addons.bal

```diff
@@ -0,0 +1,215 @@
+        FROM TimelineSubstep
+        WHERE stepId = ${oldTimelineStepId} AND substepNr = ${oldTimelineSubstepNr};`
+    );
+    var rowCount = timelineSubstepInsertResult.affectedRowCount;
+    if rowCount == () || rowCount < 1 {
+        fail error("Cloning of TimelineSubstep with (old) timeline step id " + oldTimelineStepId.toString() + " and substepNr " + oldTimelineSubstepNr.toString() + " unsuccessful!");
+    }
+
+    // find and clone associated substepData // TODO: more than one or just one, for now don't care
+    stream<record {|int id; int dataId;|}, sql:Error?> substepDataResult = experimentDB->query(`SELECT id, dataId FROM SubstepData WHERE stepId = ${oldTimelineStepId} AND substepNr=${oldTimelineSubstepNr};`);
+    record {|int id; int dataId;|}[]|error|() substepDataList = from var data in substepDataResult
+        select data;
+    if substepDataList !is error && substepDataList !is () {
+        foreach var substepData in substepDataList {
+            // TODO: deep copy
+            int clonedExperimentDataId = check cloneExperimentData(experimentDataMapping, substepData.dataId, newExperimentId, oldExperimentId);
+
+            //   clone step data with id of new experiment data
+            var stepDataInsertResult = check experimentDB->execute(
+                `INSERT INTO SubstepData (stepId, substepNr, dataId, relationType) 
+                SELECT ${newTimelineStepId}, ${oldTimelineSubstepNr}, ${clonedExperimentDataId}, relationType
+                FROM StepData
+                WHERE stepId = ${oldTimelineStepId} AND dataId = ${substepData.dataId};`
+            );
+            rowCount = stepDataInsertResult.affectedRowCount;
+            if rowCount == () || rowCount < 1 {
+                fail error("Cloning of SubstepData with (old) timeline step id " + oldTimelineStepId.toString() + " and dataId " + clonedExperimentDataId.toString() + " unsuccessful!");
+            }
+        }
+    }
+}
+
+# Clones timeline step including associated step data, experiment data, and substeps.
+#
+# + newExperimentId - experiment id of the new (cloned) experiment
+# + oldExperimentId - experiment id of the experiment that is being cloned
+# + oldTimelineStepId - step id of (old) timeline step that is to be cloned
+# + experimentDataMapping - mutable mapping of old ExperimentData item ids to new experiment data ids to check if new experiment data needs to be created or not
+# + return - The new (cloned) timeline step id or the encountered error
+public isolated transactional function cloneTimelineStepComplex(int newExperimentId, int oldExperimentId, int oldTimelineStepId, map<int> experimentDataMapping) returns ()|error {
+    // clone timeline step
+    var newTimelineStepId = check cloneTimelineStepSimple(newExperimentId, oldExperimentId, oldTimelineStepId);
+
+    // find and clone associated step data // TODO: more than one or just one, for now don't care
```

Comment:
what do these todos mean?

Reply:
Wasn't sure if there is more than one step data per step (also substep data for substeps). If there is at most one the code could be refactored a little for better understanding.

Reply:
Every step/substep can produce 0..n data outputs

Reply:
Ok thanks that clears it up.